### DISCUSSION TITLE
Something that would be nice to have, run custom code

### DISCUSSION THREAD

**Discussion by agluck91** [blinksh/blink#1734] [Ideas]
Not sure how well this can be integrated. I  know that there are many different types of ipads. I would love to see the ability to run custom code. Like a typescript interpreter or compiler. Or javascript. There are plenty of tools that let you run JS on the ipad. Would be nice to have it build into blink. 

This could introduce a lot of power to the user. I would love to use it ot trigger custom workflows. 

Or at minimum the ability to script in the cmd line similar to bash scripting. Or even allow us to easily add custom swift programs to run from a bin directory. That would work too. Lots of ways to skin this one.

**Comment by rodrigoIkari** [comment]
There is another app for developers that includes nodejs-mobile and CPython, to allow to run code locally on IPad.
I think it would be great to do something similar on Blink Shell, expanding for example what is available on the shell itself (running npm, or custom node/python scripts)

**Reply by carloscabanero** [reply]
We just introduced Blink Build that gives you a full coding environment with the default Hacker Tools. There is no limit on how far you can take it. You can use it for real Python, Node, we even built Build with Build in Rust. We also use it to push our docs to Vercel, or with the gcloud CLI.

Our goal has always been to make remote development feel local. Not just to overcome the limitations of iOS and what apps can do, but because I hate state in my machines too. But having some way to trigger workflows from Blink, maybe that would work for some scenarios.

**Comment by carloscabanero** [comment]
For coding itself we just introduced Blink Build. 

But to understand the real use case here, what custom workflows would you like to trigger from Blink? This is definitely something we are thinking a lot atm.

**Reply by agluck91** [reply]
Sure, For me, for a start, it would not have to be very complex, like introducing full bash style scripting. Your introduction of build has got me in your code base to understand what you are doing there. I love that you, in the code, are able to spin up a container instance and connect WG to it. You all blew my mind with that. So I have been trying to understand how you all accomplished that. 


I will probably use build, I am certainly going to subscribe to it just to help support development, but I already have full workflows built up with gitops and all in my local server rack. Being able to trigger some pulumi jobs in my server rack would be sufficient enough. I can probably use curl to trigger this on the ipad, and just setup some APIs that can handle it on the server side. Then have an API return the link, and ssh keys and all that good stuff. 

Just having an HA bastion host that has all the important stuff on it like pulumi, kubectl, and such is probably the answer here, but if there were something more native that could allow me to kick stuff off, or string together many commands and just have a `./start` concept it might be nice.  

I love typescript as well, and I know that there have been some people that have been able to get it to run on iPad, so having something like that integrated in the shell without leaving the iPad would be really cool, but not necessary. For that, I would just be writing automation scripts most likely. Some times you have tasks that you would like to run on a server and not have to login to the server to kick it off. If we had TS or JS support then I could almost abstract all of my kubernetes commands using the K8s API layer then in my local shell as well. Again, These are just nice to haves, it just takes one step out of my workflow, and that is to login to a bastion host. 

Seeing you dynamically spin up a container and connect to it with a fresh WG instance was awesome! Right now, I have to go out of the blink app, connect to WG, then come back in, and then mosh over to the bastion server. So instead of being able to run APIs from local, maybe you can broaden the WG scripts to be able to open our already existing wg configurations, and login to a host automatically? I was thinking after understanding your code, to open a PR for something like that. Perhaps expanding the mosh functionality?

**Reply by agluck91** [reply]
I love the idea of blink build, and I am thinking of ways to use it, but it is still limited, and I have a server rack in my house that is always running, so that is where I live most of the time. I have a machine in the rack that is dedicated for development for instance, so build does not appeal to me as much.

**Reply by agluck91** [reply]
There is also the thought of a plugin directory, where we can write out own binaries in swift and drop them in to be triggered from blink?

**Reply by Numbski** [reply]
Hah, I am over here trying to get into build, but it doesn't take show up under Blink Classic it would appear.  Only Blink+, correct?

**Reply by carloscabanero** [reply]
Yes, Build requires Blink+ functionality. You can Sign Up for Blink+Build, which includes a one week trial of Build.
